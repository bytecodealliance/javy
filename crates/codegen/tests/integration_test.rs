use std::path::PathBuf;

use anyhow::Result;
use javy_codegen::{Generator, LinkingKind, Plugin, JS};

#[test]
fn test_empty() -> Result<()> {
    // Load valid JS from file.
    let js = JS::from_file(
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("tests")
            .join("sample-scripts")
            .join("empty.js")
            .as_path(),
    )?;

    // Load existing Javy plugin.
    let plugin = Plugin::new_from_path(
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("default_plugin.wasm"),
    )?;

    // Configure code generator.
    let mut generator = Generator::new(plugin);
    generator.linking(LinkingKind::Static);

    // Generate valid Wasm module.
    let wasm = generator.generate(&js)?;

    // Verify generated bytes are valid Wasm.
    walrus::Module::from_buffer(wasm.as_slice())?;

    Ok(())
}
