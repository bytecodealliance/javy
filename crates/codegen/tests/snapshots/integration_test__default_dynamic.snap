---
source: crates/codegen/tests/integration_test.rs
expression: wat
---
(module
  (type (;0;) (func))
  (type (;1;) (func (param i32 i32 i32 i32)))
  (type (;2;) (func (param i32 i32 i32 i32) (result i32)))
  (import "javy_quickjs_provider_v4" "canonical_abi_realloc" (func (;0;) (type 2)))
  (import "javy_quickjs_provider_v4" "invoke" (func (;1;) (type 1)))
  (import "javy_quickjs_provider_v4" "memory" (memory (;0;) 0))
  (export "_start" (func 4))
  (export "log" (func 2))
  (export "log2" (func 3))
  (func (;2;) (type 0)
    (local i32 i32)
    i32.const 0
    i32.const 0
    i32.const 1
    i32.const 296
    call 0
    local.tee 0
    i32.const 0
    i32.const 296
    memory.init 0
    data.drop 0
    i32.const 0
    i32.const 0
    i32.const 1
    i32.const 3
    call 0
    local.tee 1
    i32.const 0
    i32.const 3
    memory.init 1
    data.drop 1
    local.get 0
    i32.const 296
    local.get 1
    i32.const 3
    call 1
  )
  (func (;3;) (type 0)
    (local i32 i32)
    i32.const 0
    i32.const 0
    i32.const 1
    i32.const 296
    call 0
    local.tee 0
    i32.const 0
    i32.const 296
    memory.init 0
    data.drop 0
    i32.const 0
    i32.const 0
    i32.const 1
    i32.const 4
    call 0
    local.tee 1
    i32.const 0
    i32.const 4
    memory.init 2
    data.drop 2
    local.get 0
    i32.const 296
    local.get 1
    i32.const 4
    call 1
  )
  (func (;4;) (type 0)
    (local i32)
    i32.const 0
    i32.const 0
    i32.const 1
    i32.const 296
    call 0
    local.tee 0
    i32.const 0
    i32.const 296
    memory.init 0
    local.get 0
    i32.const 296
    i32.const 0
    i32.const 0
    call 1
  )
  (data (;0;) "C\07\18function.mjs\06log\08log2\0econsole4Hello from top-level scope(Hello from function!*Hello from function2!\0d\c8\03\00\02\00\00\ca\03\00\01\cc\03\00\00\00\0c \06\01\a4\01\00\00\00\03\02\02\1f\00\ca\03\00\01\cc\03\01\01\08\ec\08\c2\00\e3\c2\01\e4)8\e7\00\00\00B\e5\00\00\00\04\e8\00\00\00$\01\00\0e\06.\c8\03\01\04\01\00\0a\12\00\09\00\0f\10\00\05\08\00\05\07\0cC\06\01\ca\03\00\00\00\03\00\00\13\008\e7\00\00\00B\e5\00\00\00\04\e9\00\00\00$\01\00)\c8\03\01\02\03]\07\0c\00\00\05\00\05\10\00\05\08\00\05\07\0cC\06\01\cc\03\00\00\00\03\00\00\13\008\e7\00\00\00B\e5\00\00\00\04\ea\00\00\00$\01\00)\c8\03\05\02\03]\07\0c\00\00\05\00\05\10\00\05\08\00\05\07")
  (data (;1;) "log")
  (data (;2;) "log2")
  (@producers
    (language "JavaScript" "ES2020")
    (processed-by "Javy" "2.0.0-alpha.1")
  )
  (@custom "javy_source" (after data) "export function log() {\0a    console.log(\22Hello from function!\22);\0a}\0a\0aexport function log2() {\0a    console.log(\22Hello from function2!\22);\0a}\0a\0aconsole.log(\22Hello from top-level scope\22);\0a")
)
